keps/sig-apps/20190318-PodDisruptionBudget-graduation-to-stable.md

```diff
@@ -0,0 +1,132 @@
+long term.
+
+### Goals
+
+* Plan to promote PDB API to stable version.
+* List action items to mitigate potential risks of supporting mutable PDB.
+
+### Non-Goals
+
+* Making changes to the API fields or their meanings.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+#### Mutable PDBs
+
+A mutable PDB object allows its `MinAvailable`, `MaxUnavailable`, and `Selector`
+fields to be modified by clients. Components that use PDB must watch such
+modifications and use the updated values when making decisions.
+
+This feature is implemented by [this PR](https://github.com/kubernetes/kubernetes/pull/69867).
+
+### Risks and Mitigations
+
+We plan to support mutation of PDB objects that didn't exist in previous versions.
+The following needs to be checked and verified before graduation of the API.
+
+- [ ] Ensure that components do not have any logic that relies on immutability
+of PDBs. For example, if component builds a cache of pods that match various
+PDBs, they must add logic to invalid the cache on updates.
+   - Action Item: sweep for in-tree uses to make sure we’re informer driven and
+   aren’t assuming immutability.
+- [ ] Check what watches PDBs and make sure there is no performance concerns.
+- [ ] Updating a PDB could cause the state of a cluster to seem incorrect. For
+example, a PDB states that at least 10 replicas of a collection must be alive.
+Kubernetes control plane evicts some of the existing pods, but keeps at least 10
+around. The PDB is updated and states that at least 20 replicas must be kept
+alive. It may appear to an observer that the evictions happened before the PDB 
+update were incorrect, if they don’t notice the PDB update.
+  - Action Item: Update documents and explain mutablility of PDB its side effects.
+
+### Test Plan
+
+#### Existing Tests
+PodDisruptionBudget currently has tests in various components that use the feature:
+
+* Disruption controller
+  - [This test](https://github.com/kubernetes/kubernetes/blob/687d759e362b05dcdf11e336e2799704918e048d/pkg/controller/disruption/disruption_test.go#L140)
+  tests PDB MinAvailable, MaxUnavailable, and selector functionality in Disruption controller.
+* Kubectl
+  - [This test](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/generate/versioned/pdb_test.go)
+  tests generation of a PDB objects out of given parameters
+  - [This test](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/create/create_pdb_test.go)
+  tests creation of PDB objects from cmd parameters
+* Scheduler
+  - [This test](https://github.com/kubernetes/kubernetes/blob/ac56bd502ab96696682c66ebdff94b6e52471aa3/test/integration/scheduler/preemption_test.go#L731)
+  tests effects of PDB on preemption (PDB is honored in a best effort way)
+* Eviction integration tests
+  - [These tests](https://github.com/kubernetes/kubernetes/blob/master/test/integration/evictions/evictions_test.go) test eviction logic and its interactions with PDB.
```

Comment:
A bunch of autoscaling tests: https://k8s-testgrid.appspot.com/sig-autoscaling-cluster-autoscaler#gci-gce-autoscaling&include-filter-by-regex=pdb, code in: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/autoscaling/cluster_size_autoscaling.go

Reply:
Added the test